contracts/Cross-chain/OmnichainGovernanceExecutor.sol

```diff
@@ -0,0 +1,275 @@
+     * @param proposalId_ Id of proposal that is to be executed.
+     * @custom:event Emits ProposalExecuted with proposal id of executed proposal.
+     */
+    function execute(uint256 proposalId_) external {
+        require(
+            queued[proposalId_],
+            "OmnichainGovernanceExecutor::execute: proposal can only be executed if it is queued"
+        );
+
+        Proposal storage proposal = proposals[proposalId_];
+
+        for (uint256 i; i < proposal.targets.length; ++i) {
+            proposalTimelocks[uint8(proposal.proposalType)].executeTransaction(
+                proposal.targets[i],
+                proposal.values[i],
+                proposal.signatures[i],
+                proposal.calldatas[i],
+                proposal.eta
+            );
+        }
+        proposal.executed = true;
+        emit ProposalExecuted(proposalId_);
+    }
+
+    /**
+     * @notice Cancels a proposal only if sender is the guardian and proposal is not executed.
+     * @param proposalId_ Id of proposal that is to be cancelled.
+     * @custom:access Sender must be the guardian.
+     * @custom:event Emits ProposalCancelled with proposal id of the cancelled proposal.
+     */
+    function cancel(uint256 proposalId_) external {
+        Proposal storage proposal = proposals[proposalId_];
+        require(!proposal.executed, "OmnichainGovernanceExecutor::cancel: cannot cancel executed proposal");
+        require(msg.sender == guardian, "OmnichainGovernanceExecutor::cancel: sender must be guardian");
+
+        proposal.cancelled = true;
+        for (uint256 i = 0; i < proposal.targets.length; i++) {
+            proposalTimelocks[proposal.proposalType].cancelTransaction(
+                proposal.targets[i],
+                proposal.values[i],
+                proposal.signatures[i],
+                proposal.calldatas[i],
+                proposal.eta
+            );
+        }
+
+        emit ProposalCancelled(proposalId_);
+    }
+
+    function _blockingLzReceive(
+        uint16 srcChainId_,
+        bytes memory srcAddress_,
+        uint64 nonce_,
+        bytes memory payload_
+    ) internal virtual override whenNotPaused {
+        bytes32 hashedPayload = keccak256(payload_);
+        uint256 gasToStoreAndEmit = 30000; // enough gas to ensure we can store the payload and emit the event
+
+        (bool success, bytes memory reason) = address(this).excessivelySafeCall(
+            gasleft() - gasToStoreAndEmit,
+            150,
+            abi.encodeWithSelector(this.nonblockingLzReceive.selector, srcChainId_, srcAddress_, nonce_, payload_)
+        );
+        // try-catch all errors/exceptions
+        if (!success) {
+            failedMessages[srcChainId_][srcAddress_][nonce_] = hashedPayload;
+            emit ProposalFailed(srcChainId_, srcAddress_, nonce_, reason); // Retrieve payload from the src side tx if needed to clear
+        }
+        proposalRecievedTimestamp = block.timestamp;
```

Comment:
Is this necessary? It isn't used and would actually only represent the timestamp of the last proposal received

Reply:
It may be useful at frontend?

Reply:
I'm going to remove it. We can get this info from the block the event was emitted in

Reply:
Instead of this , we would need the lastProposalReceived-`proposalId` for simulation purposes and to execute the proposal after it has passed the queue delay. WDYT?